massa-execution/src/vm.rs

```diff
@@ -1,31 +1,83 @@
 
 use crate::interface_impl::INTERFACE;
 use crate::sce_ledger::{SCELedger, SCELedgerChanges};
 use crate::types::{ExecutionContext, ExecutionStep, OperationSC, StepHistory};
-use crate::ExecutionConfig;
+use crate::{ExecutionError, ExecutionSettings};
 
 lazy_static::lazy_static! {
     pub(crate) static ref CONTEXT: Arc<Mutex::<ExecutionContext>> = {
-        let ledger = SCELedger::default(); // TODO Bootstrap
-        Arc::new(Mutex::new(ExecutionContext::new(ledger)))
+        let ledger = SCELedger::default();  // will be bootstrapped later
+        let ledger_at_slot = Slot::new(0, 0); // will be bootstrapped later
+        Arc::new(Mutex::new(ExecutionContext::new(ledger, ledger_at_slot)))
     };
 }
 
 pub struct VM {
-    _cfg: ExecutionConfig,
+    _cfg: ExecutionSettings,
     step_history: StepHistory,
 }
 
 impl VM {
-    pub fn new(_cfg: ExecutionConfig) -> VM {
-        VM {
-            _cfg,
-            step_history: Default::default(),
+    pub fn new(
+        cfg: ExecutionSettings,
+        thread_count: u8,
+        ledger_bootstrap: Option<(SCELedger, Slot)>,
+    ) -> Result<VM, ExecutionError> {
+        // bootstrap ledger
+        let context = CONTEXT.lock().unwrap();
+        let mut final_ledger_guard = context.ledger_step.final_ledger_slot.lock().unwrap();
+
+        if let Some((ledger_bootstrap, ledger_slot)) = ledger_bootstrap {
+            // bootstrap from snapshot
+            *final_ledger_guard = (ledger_bootstrap, ledger_slot);
+        } else {
+            // not bootstrapping: load initial SCE ledger from file
+            let ledger_slot = Slot::new(0, thread_count.saturating_sub(1)); // last genesis block
+            let ledgger_balances = serde_json::from_str::<AddressHashMap<Amount>>(
+                &std::fs::read_to_string(&cfg.initial_sce_ledger_path).map_err(|err| {
+                    ExecutionError::FileError(format!(
+                        "error loading initial SCE ledger file {}: {}",
+                        cfg.initial_sce_ledger_path
+                            .to_str()
+                            .unwrap_or("(non-utf8 path)"),
+                        err
+                    ))
+                })?,
+            )
+            .map_err(|err| {
+                ExecutionError::FileError(format!(
+                    "error parsing initial SCE ledger file {}: {}",
+                    cfg.initial_sce_ledger_path
+                        .to_str()
+                        .unwrap_or("(non-utf8 path)"),
+                    err
+                ))
+            })?;
+            let ledger_bootstrap = SCELedger::from_balances_map(ledgger_balances);
+            *final_ledger_guard = (ledger_bootstrap, ledger_slot);
         }
+
+        Ok(VM {
+            _cfg: cfg,
+            step_history: Default::default(),
+        })
+    }
+
+    // clone bootstrap state (final ledger and slot)
+    pub fn get_bootstrap_state(&self) -> (SCELedger, Slot) {
+        CONTEXT
```

Comment:
So this, and other uses of `CONTEXT`, will change with https://github.com/massalabs/sc-runtime/pull/35, and it should not be a problem.

Reply:
good